### ISSUE TITLE
KeyError when using the dmin flag

### ISSUE THREAD

**Issue by apeck12** [rs-station/careless#71]
Running careless (v0.2.3) on an mtz file of XFEL data as follows:
```
careless mono --mlp-layers=10 --image-layers=2 --dmin=2.0 --iterations=30_000 --studentt-likelihood-dof=32 "dHKL,XDET,YDET,ewald_offset" test.mtz dmin_2A
```
exited with the error:
```
Traceback (most recent call last):
  File "/cds/home/a/apeck/miniconda3/bin/careless", line 8, in <module>
    sys.exit(main())
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/careless/careless.py", line 9, in main
    run_careless(parser)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/careless/careless.py", line 30, in run_careless
    inputs,rac = df.format_files(parser.reflection_files)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/careless/io/formatter.py", line 124, in format_files
    return self((load(f) for f in files))
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/careless/io/formatter.py", line 102, in __call__
    return self.finalize(data, rac)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/careless/io/formatter.py", line 319, in finalize
    refl_id = rac.to_refl_id(
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/careless/io/asu.py", line 166, in to_refl_id
    return self.asu_and_miller_lookup_table.loc[idx, 'id'].to_numpy('int')
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 961, in __getitem__
    return self._getitem_tuple(key)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1140, in _getitem_tuple
    return self._getitem_lowerdim(tup)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 843, in _getitem_lowerdim
    return self._getitem_nested_tuple(tup)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 942, in _getitem_nested_tuple
    obj = getattr(obj, self.name)._getitem_axis(key, axis=axis)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1194, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1132, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1330, in _get_listlike_indexer
    keyarr, indexer = ax._get_indexer_strict(key, axis_name)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexes/multi.py", line 2590, in _get_indexer_strict
    return super()._get_indexer_strict(key, axis_name)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 5796, in _get_indexer_strict
    self._raise_if_missing(keyarr, indexer, axis_name)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexes/multi.py", line 2610, in _raise_if_missing
    return super()._raise_if_missing(key, indexer, axis_name)
  File "/cds/home/a/apeck/miniconda3/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 5859, in _raise_if_missing
    raise KeyError(f"{not_found} not in index")
KeyError: '[(0, 24, 17, 13), (0, 38, 6, 5), (0, 30, 22, 7), (0, 34, 19, 4), (0, 37, 3, 7), (0, 29, 5, 13), (0, 37, 8, 6), (0, 32, 23, 3), (0, 13, 6, 18), (0, 32, 17, 8), (0, 34, 18, 5), (0, 35, 18, 3), (0, 30, 13, 11), (0, 29, 26, 4), (0, 28, 23, 8), (0, 29, 18, 10), (0, 28, 17, 11), (0, 39, 2, 4), (0, 31, 25, 0), (0, 22, 12, 15), (0, 25, 2, 15), (0, 26, 22, 10), (0, 34, 2, 10), (0, 35, 16, 5), (0, 28, 9, 13), (0, 35, 19, 1), (0, 36, 9, 7), (0, 32, 7, 11), (0, 33, 22, 2), (0, 37, 4, 7), (0, 39, 8, 1), (0, 35, 19, 0), (0, 39, 1, 4), (0, 39, 0, 4), (0, 33, 20, 5), (0, 36, 17, 1), (0, 29, 20, 9), (0, 23, 21, 12), (0, 39, 8, 0), (0, 27, 5, 14), (0, 34, 15, 7), (0, 36, 17, 0), (0, 29, 27, 2), (0, 31, 4, 12), (0, 34, 3, 10), (0, 38, 12, 1), (0, 35, 4, 9), (0, 26, 14, 13), (0, 36, 16, 3), (0, 16, 10, 17), (0, 30, 25, 4), (0, 36, 15, 4), (0, 37, 15, 0), (0, 33, 15, 8), (0, 23, 10, 15), (0, 30, 26, 2), (0, 36, 12, 6), (0, 38, 9, 4), (0, 31, 25, 1), (0, 31, 17, 9), (0, 38, 12, 0), (0, 36, 5, 8), (0, 34, 20, 3), (0, 33, 17, 7), (0, 20, 10, 16), (0, 26, 20, 11), (0, 32, 19, 7), (0, 22, 4, 16), (0, 12, 8, 18), (0, 39, 7, 2), (0, 18, 6, 17), (0, 31, 19, 8), (0, 23, 15, 14), (0, 33, 9, 10), (0, 7, 0, 19), (0, 24, 20, 12), (0, 7, 1, 19), (0, 25, 3, 15), (0, 35, 5, 9), (0, 26, 9, 14), (0, 30, 9, 12), (0, 32, 12, 10), (0, 32, 15, 9), (0, 28, 14, 12), (0, 27, 12, 13)] not in index'
```
This seems to be specific to the `dmin` flag, since training finished when that argument was omitted. In case it's useful, the mtz file was generated from a stream file using the stream2mtz script with a single set of cell constants.

**Comment by kmdalton** [comment]
hi, @apeck12, can you share your `mtz` or a minimal version thereof?

**Comment by kmdalton** [comment]
This may be related #48

**Comment by apeck12** [comment]
Here's a link to the mtz: https://slac-my.sharepoint.com/:u:/g/personal/apeck_slac_stanford_edu/EVirvxzvn21GmaUnGp-4F74Bk8KMTCkCrOApytUw_DuTiA?email=kmdalton%40g.harvard.edu&e=Ct2Vmy. I converted this from a CrystFEL stream file by running:
```
stream2mtz test.stream -o test.mtz -g 96 -c 79.629 79.824 38.591 90.165 89.969 90.124
```
which seems to overwrite the individual cell constants with the supplied cell parameters.

**Comment by kmdalton** [comment]
okay -- i'll have a look and see if i can learn anything.

**Comment by apeck12** [comment]
Sorry for my oversight, and thanks for catching that. That seems to be the crux of the issue since the `ReciprocalASU` and `Dataset` objects end up having different sets of hkls in this case of inconsistent cell and space group. 

(Though as far as I could tell, the space group is only used to map reflections to the asu, and both objects have the same cell and use the cell's `calculate_d_array` function to compute resolution. So I’m still confused as to how the mismatch occurs.)

**Comment by kmdalton** [comment]
Yeah I'm not convinced I'm to the bottom of this yet. At least you have a workaround for this data though.

**Comment by JBGreisman** [comment]
Just looking at the first few Miller indices in the `KeyError`, it does seem it is looking up reflections that are higher resolution than the specified `dmin`:
```
In [37]: cell
Out[37]: <gemmi.UnitCell(79.629, 79.824, 38.591, 90.165, 89.969, 90.124)>

In [38]: rs.utils.compute_dHKL(np.array([[24, 17, 13], [38, 6, 5], [30, 22, 7]]), cell)
Out[38]: array([1.9989645, 1.9987396, 1.9941647], dtype=float32)
``` 
Does this only happen with "invalid" cells? This seems a bit different to me than #48, because only a single MTZ was provided

**Comment by kmdalton** [comment]
I don't see any reason it should be exclusive to "invalid" cells. I think it is probably more subtle than that. I have yet to reproduce the error with a valid cell yet, but I will continue tinkering.

**Comment by kmdalton** [comment]
Okay -- I think the issue does have to do with the cell being improper. I apply the resolution cutoff in the `careless` parser before mapping reflections to reciprocal ASU. In the case of an improper cell, the resolution of the reflections change when you map the into the ASU. So, there is a population of reflections within the resolution cutoff that get changed to be outside the resolution range. There are a couple of options for dealing with this. 
 1) Always compute `dHKL` in the reciprocal ASU. This can probably be done in a way that won't add a crazy amount of overhead, but I'm not sure yet. I really would like to only call `hkl_to_asu` one time if possible, and it's certainly a bad idea to duplicate the dataset. I'm sure this is solvable, but I might need to think a bit. 
 2) We could just say we're not accepting improper cells and throw a helpful error message. 
 
 Any thoughts from users? @apeck12 ?

**Comment by apeck12** [comment]
Ah, got it. Thanks again for looking into this. I’d be in favor of the second option, since I should have caught the cell/space group mismatch while converting the stream file instead of using the average cell.

**Comment by JBGreisman** [comment]
I agree with raising a helpful error message in this case. Since `careless` just passes the input cell forward to the merged MTZ, I feel like its a good opportunity to raise an error. 

This sort of check is straightforward to do with `gemmi`:
```python
In [1]: import gemmi

In [2]: cell = gemmi.UnitCell(79.629, 79.824, 38.591, 90.165, 89.969, 90.124)

In [3]: cell.is_compatible_with_spacegroup(gemmi.SpaceGroup(96))
Out[3]: False

In [4]: cell.is_compatible_with_spacegroup(gemmi.SpaceGroup(1))
Out[4]: True
```

**Comment by kmdalton** [comment]
okay -- i will put an error in before the next release. thanks everyone

**Comment by kmdalton** [comment]
resolved in #172
